### DISCUSSION TITLE
Typing Improvement - Client mounts

### DISCUSSION THREAD

**Discussion by tomtom103** [encode/httpx#2979] [Potential Issue] [answered]
The `httpx.Client` defines the `self._mounts` as `Dict[URLPattern, typing.Optional[BaseTransport]]`, but the constructor argument is typed as: `Optional[Mapping[str, BaseTransport]]`.

Following the logic described in `https://www.python-httpx.org/advanced/#no-proxy-support`, when manually providing mounts to the client (for example, disabling certain features for a specific domain), we should be able to provide a `None` transport to ensure the transport used will be the one provided by default to the client instead of matching with one of the mounts.

Take the following example:

```python
import httpx

mounts = {
    "http://": httpx.HTTPTransport(proxy=httpx.Proxy(url="http://my-proxy:8030")),
    "https://": httpx.HTTPTransport(proxy=httpx.Proxy(url="http://my-proxy:8030")),
    "all://*.my-domain.com": None # Disable all proxying to specific domain and use specified transport instead
}

with httpx.Client(
    transport=httpx.HTTPTransport(verify=False),
    mounts=mounts
) as client:
    # Should not use a proxy
    # Defaults to the provided transprot
    response = client.get("https://foo.my-domain.com/get")

    # Should use a proxy
    response = client.get("https://httpbin.org/get")
```

Type-checker fails with:
```
Argument "mounts" to "Client" has incompatible type "Dict[str, Optional[HTTPTransport]]"; expected "Optional[Mapping[str, BaseTransport]]"
```

But seems to be a valid use-case here according to the logic defined in the `Client._transport_for_url`. This is also a valid use-case at  runtime, I'm wondering if this is something the types should also support?

**Comment by karpetrosyan** [comment, marked as answer]
I think it has been resolved in https://github.com/encode/httpx/pull/2879
